File: pipeline/analyze/extract_drinking_windows.py
#!/usr/bin/env python3
"""
Extracts drinking window data from Ridge vintage notes in JSONL,
then updates wine_vintages with producer_drinking_window_start/end.

Usage:
    python -m pipeline.analyze.extract_drinking_windows          # Dry run
    python -m pipeline.analyze.extract_drinking_windows --apply  # Update DB
"""

import argparse
import json
import re
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[2]))

from pipeline.lib.db import get_supabase

JSONL_FILE = Path(__file__).resolve().parents[2] / "ridge_wines.jsonl"

NUMBER_WORDS = {
    "one": 1, "two": 2, "three": 3, "four": 4, "five": 5, "six": 6,
    "seven": 7, "eight": 8, "nine": 9, "ten": 10, "eleven": 11,
    "twelve": 12, "thirteen": 13, "fourteen": 14, "fifteen": 15,
    "sixteen": 16, "seventeen": 17, "eighteen": 18, "nineteen": 19,
    "twenty": 20, "twenty-one": 21, "twenty-two": 22, "twenty-three": 23,
    "twenty-four": 24, "twenty-five": 25, "thirty": 30, "thirty-five": 35,
    "forty": 40, "fifty": 50,
}


def parse_word(w: str) -> int | None:
    n = NUMBER_WORDS.get(w.lower())
    if n:
        return n
    try:
        return int(w)
    except ValueError:
        return None


def extract_drinking_window(notes: str, vintage: int) -> dict | None:
    if not notes or not vintage:
        return None

    patterns = [
        re.compile(r"(?:over|for)\s+the\s+next\s+([\w-]+)\s+(?:to\s+([\w-]+)\s+)?years?", re.I),
        re.compile(r"(?:develop|improve|evolve|age)\s+(?:over|for)\s+(?:the\s+next\s+)?([\w-]+)\s+(?:to\s+([\w-]+)\s+)?years?", re.I),
        re.compile(r"enjoy\s+(?:now\s+and\s+)?(?:over|for)\s+(?:the\s+next\s+)?([\w-]+)\s+(?:to\s+([\w-]+)\s+)?years?", re.I),
        re.compile(r"drink\s+now\s+and\s+(?:over|for)\s+(?:the\s+next\s+)?([\w-]+)\s+(?:to\s+([\w-]+)\s+)?years?", re.I),
    ]
    for pat in patterns:
        m = pat.search(notes)
        if m:
            hi = parse_word(m.group(2)) if m.group(2) else parse_word(m.group(1))
            if hi:
                return {"start": vintage, "end": vintage + hi, "source": m.group(0)}

    # "best from YYYY"
    m = re.search(r"best\s+(?:from|after)\s+(\d{4})", notes, re.I)
    if m:
        return {"start": int(m.group(1)), "end": int(m.group(1)) + 15, "source": m.group(0)}

    # "YYYY-YYYY" range
    m = re.search(r"(\d{4})\s*[-\u2013]\s*(\d{4})", notes)
    if m:
        s, e = int(m.group(1)), int(m.group(2))
        if 1990 <= s <= 2050 and e >= s and e <= 2100:
            return {"start": s, "end": e, "source": m.group(0)}

    return None


def main():
    parser = argparse.ArgumentParser(description="Extract drinking windows from Ridge JSONL")
    parser.add_argument("--apply", action="store_true", help="Actually update DB")
    args = parser.parse_args()

    sb = get_supabase()
    print(f"Mode: {'APPLY (will update DB)' if args.apply else 'DRY RUN (preview only)'}\n")

    lines = JSONL_FILE.read_text(encoding="utf-8").strip().split("\n")
    wines = [json.loads(l) for l in lines]
    print(f"Loaded {len(wines)} entries from JSONL\n")

    result = sb.table("producers").select("id").eq("slug", "ridge-vineyards").execute()
    if not result.data:
        print("Ridge producer not found")
        sys.exit(1)

    db_wines = sb.table("wines").select("id,name").eq("producer_id", result.data[0]["id"]).execute()
    wine_name_to_id = {w["name"]: w["id"] for w in (db_wines.data or [])}

    extracted = 0
    updated = 0
    skipped = 0

    for w in wines:
        if not w.get("vintage") or not w.get("wineName"):
            continue
        notes = f"{w.get('vintageNotes', '')} {w.get('winemakerNotes', '')}"
        dw = extract_drinking_window(notes, w["vintage"])
        if not dw:
            continue
        extracted += 1

        wine_id = wine_name_to_id.get(w["wineName"])
        if not wine_id:
            print(f"  No DB wine for \"{w['wineName']}\"")
            skipped += 1
            continue

        print(f"  {w['vintage']} {w['wineName']}: {dw['start']}-{dw['end']} (from: \"{dw['source']}\")")

        if args.apply:
            result = sb.table("wine_vintages").update({
                "producer_drinking_window_start": dw["start"],
                "producer_drinking_window_end": dw["end"],
            }).eq("wine_id", wine_id).eq("vintage_year", w["vintage"]).execute()
            if result.data is not None:
                updated += 1
            else:
                print("    UPDATE ERROR")

    print(f"\n{'=' * 40}")
    print("  Drinking Window Extraction")
    print("=" * 40)
    print(f"  Scanned: {len(wines)} entries")
    print(f"  Extracted: {extracted}")
    print(f"  Skipped: {skipped}")
    if args.apply:
        print(f"  Updated in DB: {updated}")
    else:
        print("  (Dry run -- use --apply to update DB)")


if __name__ == "__main__":
    main()

File: scripts_archive/node/root/extract_drinking_windows.mjs
#!/usr/bin/env node
/**
 * extract_drinking_windows.mjs
 *
 * Extracts drinking window data from Ridge vintage notes already in the JSONL,
 * then updates the wine_vintages table with producer_drinking_window_start/end.
 *
 * Usage:
 *   node extract_drinking_windows.mjs          # Dry run (show what would update)
 *   node extract_drinking_windows.mjs --apply  # Actually update DB
 */

import { readFileSync } from 'fs';
import { createClient } from '@supabase/supabase-js';

// ── Load .env ───────────────────────────────────────────────
const envPath = new URL('.env', import.meta.url).pathname.replace(/^\/([A-Z]:)/, '$1');
const envContent = readFileSync(envPath, 'utf-8');
for (const line of envContent.split('\n')) {
  const trimmed = line.replace(/\r/g, '').trim();
  if (!trimmed || trimmed.startsWith('#')) continue;
  const eqIdx = trimmed.indexOf('=');
  if (eqIdx === -1) continue;
  const key = trimmed.slice(0, eqIdx);
  const val = trimmed.slice(eqIdx + 1);
  if (!process.env[key]) process.env[key] = val;
}

const sb = createClient(process.env.SUPABASE_URL, process.env.SUPABASE_SERVICE_ROLE || process.env.SUPABASE_ANON_KEY);
const APPLY = process.argv.includes('--apply');

const NUMBER_WORDS = {
  'one':1,'two':2,'three':3,'four':4,'five':5,'six':6,'seven':7,'eight':8,'nine':9,'ten':10,
  'eleven':11,'twelve':12,'thirteen':13,'fourteen':14,'fifteen':15,'sixteen':16,'seventeen':17,
  'eighteen':18,'nineteen':19,'twenty':20,'twenty-one':21,'twenty-two':22,'twenty-three':23,
  'twenty-four':24,'twenty-five':25,'twenty-six':26,'twenty-seven':27,'twenty-eight':28,
  'twenty-nine':29,'thirty':30,'thirty-five':35,'forty':40,'fifty':50,
};

function parseWord(w) {
  return NUMBER_WORDS[w.toLowerCase()] || parseInt(w) || null;
}

function extractDrinkingWindow(notes, vintage) {
  if (!notes || !vintage) return null;

  // Pattern 1: "over/for the next X (to Y) years"
  let m = notes.match(/(?:over|for)\s+the\s+next\s+([\w-]+)\s+(?:to\s+([\w-]+)\s+)?years?/i);
  if (m) {
    const hi = m[2] ? parseWord(m[2]) : parseWord(m[1]);
    if (hi) return { start: vintage, end: vintage + hi, source: m[0] };
  }

  // Pattern 2: "develop/improve/evolve over/for (the next) X (to Y) years"
  m = notes.match(/(?:develop|improve|evolve|age)\s+(?:over|for)\s+(?:the\s+next\s+)?([\w-]+)\s+(?:to\s+([\w-]+)\s+)?years?/i);
  if (m) {
    const hi = m[2] ? parseWord(m[2]) : parseWord(m[1]);
    if (hi) return { start: vintage, end: vintage + hi, source: m[0] };
  }

  // Pattern 3: "enjoy now and over the next X years"
  m = notes.match(/enjoy\s+(?:now\s+and\s+)?(?:over|for)\s+(?:the\s+next\s+)?([\w-]+)\s+(?:to\s+([\w-]+)\s+)?years?/i);
  if (m) {
    const hi = m[2] ? parseWord(m[2]) : parseWord(m[1]);
    if (hi) return { start: vintage, end: vintage + hi, source: m[0] };
  }

  // Pattern 4: "drink now and over the next X years" or "drink now and for the next X years"
  m = notes.match(/drink\s+now\s+and\s+(?:over|for)\s+(?:the\s+next\s+)?([\w-]+)\s+(?:to\s+([\w-]+)\s+)?years?/i);
  if (m) {
    const hi = m[2] ? parseWord(m[2]) : parseWord(m[1]);
    if (hi) return { start: vintage, end: vintage + hi, source: m[0] };
  }

  // Pattern 5: "best from YYYY" or "best after YYYY"
  m = notes.match(/best\s+(?:from|after)\s+(\d{4})/i);
  if (m) {
    return { start: parseInt(m[1]), end: parseInt(m[1]) + 15, source: m[0] };
  }

  // Pattern 6: "YYYY-YYYY" or "YYYY–YYYY" drinking window range
  m = notes.match(/(\d{4})\s*[-–]\s*(\d{4})/);
  if (m) {
    const s = parseInt(m[1]), e = parseInt(m[2]);
    if (s >= 1990 && s <= 2050 && e >= s && e <= 2100) {
      return { start: s, end: e, source: m[0] };
    }
  }

  return null;
}

async function main() {
  console.log(`Mode: ${APPLY ? 'APPLY (will update DB)' : 'DRY RUN (preview only)'}\n`);

  // Load JSONL
  const lines = readFileSync('ridge_wines.jsonl', 'utf8').trim().split('\n');
  const wines = lines.map(l => JSON.parse(l));
  console.log(`Loaded ${wines.length} entries from JSONL\n`);

  // Get Ridge producer + wine IDs
  const { data: prod } = await sb.from('producers').select('id').eq('slug', 'ridge-vineyards').single();
  if (!prod) { console.error('Ridge producer not found'); process.exit(1); }

  const { data: dbWines } = await sb.from('wines').select('id,name').eq('producer_id', prod.id);
  const wineNameToId = new Map(dbWines.map(w => [w.name, w.id]));

  let extracted = 0;
  let updated = 0;
  let skipped = 0;

  for (const w of wines) {
    if (!w.vintage || !w.wineName) continue;

    const notes = (w.vintageNotes || '') + ' ' + (w.winemakerNotes || '');
    const dw = extractDrinkingWindow(notes, w.vintage);

    if (!dw) continue;
    extracted++;

    const wineId = wineNameToId.get(w.wineName);
    if (!wineId) {
      console.warn(`  No DB wine for "${w.wineName}"`);
      skipped++;
      continue;
    }

    console.log(`  ${w.vintage} ${w.wineName}: ${dw.start}–${dw.end} (from: "${dw.source}")`);

    if (APPLY) {
      const { error } = await sb.from('wine_vintages')
        .update({
          producer_drinking_window_start: dw.start,
          producer_drinking_window_end: dw.end,
        })
        .eq('wine_id', wineId)
        .eq('vintage_year', w.vintage);

      if (error) {
        console.error(`    UPDATE ERROR: ${error.message}`);
      } else {
        updated++;
      }
    }
  }

  console.log(`\n========================================`);
  console.log(`  Drinking Window Extraction`);
  console.log(`========================================`);
  console.log(`  Scanned: ${wines.length} entries`);
  console.log(`  Extracted: ${extracted}`);
  console.log(`  Skipped: ${skipped}`);
  if (APPLY) console.log(`  Updated in DB: ${updated}`);
  else console.log(`  (Dry run — use --apply to update DB)`);
}

main().catch(console.error);
